### DISCUSSION TITLE
How to DRYup commonly-used extension annotations?

### DISCUSSION THREAD

**Discussion by grepory** [microsoft/typespec#3903] [Q&A] [answered]
Hey folks. I'm working on hooking up our TypeSpec -> OpenAPI pipeline to generating SDKs with Speakeasy. As part of that, we have to add extension annotations to all of our endpoints that respond with paginated results, e.g.:

```
            // All operations must have a summary.
            @summary("List all alert definitions")
            @extension("x-speakeasy-pagination", {
                type: "url",
                outputs: {
                  nextUrl: "$.pageInfo.next",
                }
            })
            op list(@query skipToken?: string, @query pageSize?: int32): {
                @statusCode _: 200;
                @bodyRoot alertDefinitions: {
                    alertDefinitions: Definition[];
                    pageInfo?: PageInfo;
                };
            };
```

I'd like to make it so that teams don't have to remember or copypasta this extension annotation and others like it.

I know that I can at least replace "x-speakeasy-pagination" with a string constant, but because object constants are typed, I can't pass one in here as the second argument to extension.

```
Argument of type '#{type: "url", ref: "$.pageInfo.next"}' is not assignable to parameter of type 'unknown'TypeSpec(invalid-argument)
```

I was actually sort of surprised by this, but I am new to TypeScript as well.

I'd appreciate any guidance here. Thank you.

**Comment by grepory** [comment, marked as answer]
I think the answer here is to create a custom decorator that does this for me. :)

**Comment by timotheeguerin** [comment]
Yeah the `@extension` decorator as most was created before values and we haven't decided on how to migrate to it without breaking user.

You can however still use an alias today to reuse that [Example](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaSI7Cgp1c2luZyBPcGVuQVBJOwoKYWxpYXMgUGFnaW5hdGlvbkNvbmZpZyA9IHsKICDEQTogInVybCI7CiAgb3V0cHV0czrFHCAgbmV4dFVybDogIiQucGFnZUluZm8uxBXFLX07Cn07CgpAZXh0ZW5zaW9uKCJ4LXNwZWFrZWFzeS1wyXIiLNF%2FKQpvcCBsaXN0KCk6IHZvaWQ7Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D) 
```ts
import "@typespec/openapi";

using OpenAPI;

alias PaginationConfig = {
  type: "url";
  outputs: {
    nextUrl: "$.pageInfo.next";
  };
};

@extension("x-speakeasy-pagination", PaginationConfig)
op list(): void;
```

but yeah a custom decorator could also be a great solution and even cleaner if you really always have the same and just need to do `@myPaginated`
